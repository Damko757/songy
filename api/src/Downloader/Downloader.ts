import ffmpeg from "fluent-ffmpeg";
import { PassThrough, Readable } from "stream";
import { Metadator } from "../Metadata/Metadator.js";
import type { YTDL_DownloadOptions } from "@ybd-project/ytdl-core";
import type { Metadata } from "../../../shared/Entities/Metadata/Metadata.js";

export class Downloader {
  link: string;
  metadator: Metadator;

  constructor(link: string) {
    this.link = link;
    this.metadator = new Metadator(this.link);
  }

  audioStream(
    options: YTDL_DownloadOptions & {
      bitrate?: number;
    } = {}
  ) {
    const self = this;
    const outStream = new PassThrough();

    this.metadator
      .rawMetaData()
      .then((info) => {
        self.metadator
          .getYtDlInstance()
          .downloadFromInfo(info, { quality: "highestaudio", ...options })
          .then((stream) => {
            const reader = stream.getReader();
            ffmpeg()
              .input(
                new Readable({
                  async read() {
                    // Function to read the chunks asynchronously
                    const pushChunk = async () => {
                      try {
                        const { done, value } = await reader.read();

                        if (done) {
                          this.push(null); // No more data, signal end of stream
                        } else {
                          this.push(value); // Push the chunk of data to the Node.js stream
                          pushChunk(); // Continue reading
                        }
                      } catch (err) {
                        this.emit("error", err); // Handle any errors
                      }
                    };

                    pushChunk(); // Start reading the chunks
                  },
                })
              )
              .audioBitrate(options.bitrate ?? 320)
              .format("mp3")
              .pipe(outStream);
          })
          .catch((e) => {
            console.error(e);
            outStream.emit("error", e);
          });
      })
      .catch((e) => {
        outStream.emit("error", e);
      });

    return outStream;
  }
}
