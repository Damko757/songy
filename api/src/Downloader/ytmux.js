
// require all the things!
const fs = require("fs");
const ytdl = require('@distube/ytdl-core');
const ffmpegPath = require('ffmpeg-static');
const cp = require('child_process');
const stream = require('stream');

// default export: the ffmpeg muxer
const ytmux = (link, options = {}) => {
    const result = new stream.PassThrough({ highWaterMark: options.highWaterMark || 1024 * 512 });
    console.info("Loading:", link);
    ytdl.getInfo(link).then(info => {
      console.info(info)
      ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options }).pipe(fs.createWriteStream("out/vidOnly.mp4"));
      ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options }).pipe(fs.createWriteStream("out/audioOnly.mp4"));
      //ffmpeg -fflags +igndts -i video.mp4 -i audio.mp4 -c:v copy -c:a aac output.mp4 
        audioStream = ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options });
        videoStream = ytdl.downloadFromInfo(info, { quality: 'highestvideo', ...options });
        videoStream.on('progress', (...args) => result.emit('progress', ...args));
        videoStream.on("finish", (...args) => console.log("FINISH", ...args));
        audioStream.on("finish", (...args) => console.log("FINISH", ...args));
        // create the ffmpeg process for muxing
        ffmpegProcess = cp.spawn(ffmpegPath, [
            // supress non-crucial messages
            // '-loglevel', '8', '-hide_banner',
            "-fflags", "+igndts",
            // input audio and video by pipe
            '-i', 'pipe:3', //'-i', 'pipe:4',
            // map audio and video correspondingly
            '-map', '0:a', //'-map', '1:v',
            // no need to change the codec
            '-c', 'copy',
            // output mp4 and pipe
            '-f', 'matroska', 'pipe:5'
        ], {
            // no popup window for Windows users
            windowsHide: true,
            stdio: [
                // silence stdin/out, forward stderr,
                'inherit', 'inherit', 'inherit',
                // and pipe audio, video, output
                'pipe', 'pipe', 'pipe'
            ]
        });
        audioStream.pipe(ffmpegProcess.stdio[3]);
        videoStream.pipe(ffmpegProcess.stdio[4]);
        // ffmpegProcess.stdio[5].pipe(result);
        ffmpegProcess.stdio[5].pipe(fs.createWriteStream("out/full.mp4"));
    }).catch(err => result.emit('error', err));
    return result;
};

// export it
module.exports = ytmux;

// export other functions, in case you want them
ytmux.download = ytdl;
ytmux.chooseFormat = ytdl.chooseFormat;
ytmux.downloadFromInfo = ytdl.downloadFromInfo;
ytmux.filterFormats = ytdl.filterFormats;
ytmux.getBasicInfo = ytdl.getBasicInfo;
ytmux.getInfo = ytdl.getInfo;
ytmux.getURLVideoID = ytdl.getURLVideoID;
ytmux.getVideoID = ytdl.getVideoID;
ytmux.validateID = ytdl.validateID;
ytmux.validateURL = ytdl.validateURL;