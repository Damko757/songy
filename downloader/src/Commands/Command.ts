import type ytdl from "@distube/ytdl-core";
import type { ObjectId } from "mongoose";
import { DestroyT } from "../Workers/WorkerPool.js";

/// Worker communication commands ///
interface DownloadJob {
  id: ObjectId; ///< MongoDB entry ID
  link: string; ///< Yt link or id
  type: "audio" | "video" | "video-only"; ///< Type of file to download
  options: ytdl.downloadOptions; ///< Download options for ytdl
}

/**
 * Job for (Download) Worker
 */
export type WorkerJobInstruction =
  | ({ action: "download" } & DownloadJob)
  | { action: "ping"; value: any };

/**
 * Object defining behaviour of communication with worker
 */
export interface WorkerJob {
  job: WorkerJobInstruction; ///< Data that will get Worker
  handlers: {
    [event in WorkerMessage as event["type"]]?: (msg: event) => void;
  };
}

/**
 * Message from Worker to parent port about state of download
 */
export type WorkerMessage =
  | {
      type: "progress"; ///< Downloading file. If downloaded == total, FFMPEG conversion is underway
      downloaded: number;
      total: number;
    }
  | {
      type: "end"; ///< Finished job. Can go idle
    }
  | {
      type: "pong"; ///< Answer to ping with provided value
      value: any;
    };

/// WS communication commands ///

/**
 * Command keywords for Downloader to execute/process
 */
export enum DownloadCommandType {
  START = "DOWNLOADER_START", ///< Starts downloader and spawns its workers
  EXIT = "DOWNLOADER_EXIT", ///< Kills/Ends downloader after all workers ended downloading (or force)
  DOWNLOAD = "DOWNLOADER_DOWNLOAD", ///< Assigns new worker and starts downloading request
}
/**
 * Download command for Downloader with payload type for specific command
 */
export type DownloadCommand =
  | {
      action: DownloadCommandType.START;
      numberOfWorkers: number; ///< Size of worker pool (specific number needs to be benched)
    }
  | {
      action: DownloadCommandType.EXIT;
      force?: Exclude<DestroyT, "none">;
    }
  | ({
      action: DownloadCommandType.DOWNLOAD;
    } & DownloadJob);

/**
 * Message from Command Processor to client (API)
 */
export type DownloadMessage = {};
