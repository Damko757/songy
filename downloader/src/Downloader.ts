import { PassThrough, Readable, Writable } from "stream";
import fs from "fs";
import cp, { exec, spawn } from "child_process";
import ytdl, { type downloadOptions } from "@distube/ytdl-core";
import { ObjectId } from "mongoose";
import path, { dirname } from "path";
import { fileURLToPath } from "url";
import { DownloadJob } from "./Commands/Command.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export class Downloader {
  link: string;

  static supportedExtensions: Record<"audio" | "video", string[]> = {
    audio: ["mp3"],
    video: ["mp4"],
  };
  static downloadDirectory = path.resolve(
    __dirname,
    "..",
    "..",
    "..",
    "..",
    "out"
  );

  constructor(link: string) {
    this.link = link;
  }

  /**
   * Resolves download path
   * @param id Id of file to download
   * @param extension Extension of file. Can be undefined, then no extension is added
   * @param extra Additional value, if temporary file is created (e. g. audio and video-only files, then merge)
   */
  static downloadPath(
    id: ObjectId | string,
    extension?: DownloadJob["extension"],
    extra?: string
  ) {
    if (!fs.existsSync(Downloader.downloadDirectory))
      fs.mkdirSync(Downloader.downloadDirectory);

    // TODO: Fix for special directory
    return path.resolve(
      Downloader.downloadDirectory,
      `${id}${extra ? "_" + extra : ""}${extension ? "." + extension : ""}`
    );
  }

  /**
   * Fetches info for `this.link`
   * @param options Get info options for ytdl-core
   * @returns Information for provided link
   */
  getInfo(options: ytdl.getInfoOptions = {}) {
    return ytdl.getInfo(this.link, options);
  }

  /**
   * Validates Youtube URL/video ID
   * @returns bool
   */
  isValid() {
    return this.link.includes("?")
      ? ytdl.validateURL(this.link)
      : ytdl.validateID(this.link);
  }

  /**
   * Streams audio only
   * Original stream is converted via FFMPEG. Progress and info event are added as an extra
   * @param options
   * @returns PassThrough audio stream
   */
  audioStream(options: ytdl.downloadOptions & { bitrate?: number } = {}) {
    return new Promise<Readable>(async (resolve, reject) => {
      ytdl
        .getInfo(this.link)
        .then((info) => {
          const format =
            options.format ??
            ytdl.chooseFormat(info.formats, {
              quality: options.quality ?? "highestaudio",
              ...options,
            }); // Fetching format for FFMPEG

          const videoStream = ytdl.downloadFromInfo(info, {
            quality: options.quality ?? "highestaudio",
            filter: "audio",
            ...options,
          });

          const audioStream = new PassThrough();

          // Emitting Ytdl-core events
          videoStream.on("progress", (...args) =>
            audioStream.emit("progress", ...args)
          );
          videoStream.on("info", (...args) => {
            audioStream.emit("info", ...args);
          });

          const ffmpeg = spawn("ffmpeg", [
            "-i",
            "pipe:0", // Input from stdin
            "-vn", // No video
            ...(format.audioBitrate
              ? ["-b:a", format.audioBitrate.toString() + "k"]
              : []), // Audio bitrate from format
            "-acodec",
            "libmp3lame", // Audio codec
            "-f",
            "mp3", // Output format
            "pipe:1", // Output to stdout
          ]);

          // Pipe video input into FFmpeg
          videoStream.pipe(ffmpeg.stdin);

          // Pipe FFmpeg output to audioOutput stream
          ffmpeg.stdout.pipe(audioStream);

          ffmpeg.on("data", console.error);

          resolve(audioStream);
        })
        .catch(reject);
    });
  }

  /**
   * Download video Only
   * @param options
   */
  videoStream(options: downloadOptions = {}) {
    return new Promise<Readable>((resolve, reject) => {
      ytdl
        .getInfo(this.link)
        .then((info) => {
          const videoStream = ytdl.downloadFromInfo(info, {
            filter: "video",
            ...options,
          });
          return resolve(videoStream);
        })
        .catch(reject);
    });
  }

  /**
   * Saves `stream` into file `filename`
   * @param stream Readable stream to save
   * @param filename Of saved stream
   * @returns Promise resolved with filename upon completed write
   */
  async saveStream(stream: Readable, filename: string) {
    return new Promise<string>((resolve, reject) => {
      console.info("Saving to:", filename);
      const fileStream = fs.createWriteStream(filename);
      fileStream.on("error", reject);
      fileStream.on("close", () => resolve(filename));
      stream.pipe(fileStream);
    });
  }

  /**
   * Combines video and audio files into single MP4 file
   * @param videoFilename Filename of video file
   * @param audioFilename Filename of audio file
   * @param outFilename Filename of resulting video file
   * @returns Promise resolved upon completed combining
   */
  createCombinedVideoAudio(
    videoFilename: string,
    audioFilename: string,
    outFilename: string
  ) {
    return new Promise<void>((resolve, reject) => {
      const cmd = `ffmpeg -i ${videoFilename} -i ${audioFilename} -c:v copy -c:a aac -shortest ${outFilename} -y`;

      exec(cmd, (error, stdout, stderr) => {
        if (error) {
          console.error(`Error: ${error.message}`);
          reject(error);
          return;
        }
        if (stderr) {
          console.error(`FFmpeg stderr: ${stderr}`);
        }
        console.log("FFmpeg output:", stdout);
        resolve();
      });
    });
  }
}
