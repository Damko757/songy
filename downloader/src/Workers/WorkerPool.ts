import path from "node:path";
import { Worker } from "node:worker_threads";
import chalk, { chalkStderr } from "chalk";
import { WorkerJob, WorkerMessage } from "../Commands/Command.js";
import { Dequeue } from "../../../api/src/Helpers/Dequeue.js";

import { fileURLToPath } from "url";
import { dirname } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export type DestroyT =
  (typeof WorkerPool.DestroyType)[keyof typeof WorkerPool.DestroyType];

/**
 * Creates pool of workers,
 * each worker can be assigned job from queued jobs. Each worker's event it sent via job.handler
 */
export class WorkerPool {
  static DestroyType = {
    NONE: "none", ///< None - No destroying happening
    FORCE: "force", ///< Force - Imidiate destroying
    FINISH_ALL: "finish-all", ///< Finish-all - Waits for all planned jobs to be finished. No new jobs are scheduled
    FINISH_STARTED: "finish-started", ///< Finish-started - Waits for all started jobs to be finished. Planned jobs are discarded and no new jobs are scheduled
  } as const;

  size: number; ///< Size of workers pool

  protected waitingJobs = new Dequeue<WorkerJob>(); ///< Queue of jobs to run by workers
  protected workers = [] as Worker[]; ///< All workers
  protected idleWorkers = new Dequeue<number>(); ///< Index of idle workers
  protected assignedJobs = new Map<number, WorkerJob>(); ///< WorkerID: WorkerJob

  protected onDestroyCallback: undefined | (() => void) = undefined; ///< If not undefined, destroying is scheduled
  protected destroyType: DestroyT = WorkerPool.DestroyType.NONE;

  constructor(numberOfWorkers: number) {
    if (numberOfWorkers <= 0)
      throw new Error(`Invalid number of workers: ${numberOfWorkers}`);

    this.size = numberOfWorkers;

    this.spawnWorkers(); // Initializing workers
  }

  /**
   * Spawns assigned number of workers
   */
  protected spawnWorkers() {
    if (this.workers.length != 0)
      throw new Error("`spawnWorkers` has been called multiple times");

    for (let i = 0; i < this.size; i++) {
      const worker = new Worker(this.workerScriptFile());

      this.bindWorker(worker, i); // Binding events
      this.workers.push(worker); // Pushing to worker storage
    }
  }

  /**
   * @returns Path for worker's script file
   */
  protected workerScriptFile() {
    return path.resolve(__dirname, "Worker.js");
  }

  /**
   * Binds events to worker
   * @param worker To bind
   */
  protected bindWorker(worker: Worker, workerID: number) {
    worker.on("online", () => {
      console.log(
        chalk.greenBright(`Worker ${chalk.bold.underline(workerID)} is online!`)
      );
      this.idleWorkers.addBack(workerID); // Worker is idle
      this.workerFinished(); // Maybe it should go to work?
    });
    worker.on("message", (message: WorkerMessage) => {
      const job = this.assignedJobs.get(workerID)!;
      const handle = job.handlers[message.type];

      if (handle != undefined) {
        handle(message as any); // Specific handle will get message to process
      }

      if (message.type == "end") {
        this.workerFinished(workerID);
      }
    });
    worker.on("exit", (exitCode) => {
      console.error(
        `Worker ${chalkStderr.bold.underline(
          workerID
        )} exited with code ${chalkStderr.yellow(exitCode)}`
      );
    });
    worker.on("error", console.error);
  }

  /**
   * Called upon worker finishing its job
   * @param workerID ID of worked which ended, undefined if no worker ended, only destroying or dispatching should happen
   */
  protected workerFinished(workerID?: number) {
    if (workerID !== undefined) {
      this.idleWorkers.addBack(workerID); // Worker finished its job
      this.assignedJobs.delete(workerID); // Removing from assignation map
    }

    let dispatchedWorkerID: number | undefined = undefined; // Dispatched worker. If undefined, try to destroy
    if (this.destroyType == "none" || this.destroyType == "finish-all") {
      dispatchedWorkerID = this.dispatchWorker();
    }

    // Should not destroy
    if (this.destroyType == "none" || dispatchedWorkerID !== undefined) return;

    // Not all workers are idle
    if (this.idleWorkers.size != this.size) return;

    // Immidiate worker termination
    Promise.all(this.workers.map((worker) => worker.terminate())).then(() =>
      this.onDestroyCallback?.()
    );
  }

  /**
   * Dispatches idle worker
   * @returns ID of worker which has been dispatched or undefined if none idle or no job
   */
  protected dispatchWorker(): number | undefined {
    // Default condition for undefined
    if (!this.idleWorkers.size || !this.waitingJobs.size) return undefined;

    const job = this.waitingJobs.removeFront()!;
    const workerID = this.idleWorkers.removeFront()!;

    this.assignedJobs.set(workerID, job);
    this.workers[workerID].postMessage(job.job);

    return workerID;
  }

  /**
   * Waits for workers to finish and destroys all workers neatly
   * If `destroyType` is `"force"`, the workers are destroyed immidiately
   * @see WorkerPool.DestroyType
   * @returns Promise resolved when all workers are killed
   */
  destroy(destroyType: DestroyT) {
    return new Promise<void>((resolve, reject) => {
      this.destroyType = destroyType;
      this.onDestroyCallback = () => {
        this.onDestroyCallback = undefined; // Destroying callback
        resolve();
      };

      if (destroyType == "none") {
        // Nothing happens
        this.onDestroyCallback();
        return;
      } else if (destroyType == "force") {
        // Immidiate worker termination
        Promise.all(this.workers.map((worker) => worker.terminate()))
          .then(() => this.onDestroyCallback?.())
          .catch(reject);
        return;
      }

      this.workerFinished();
    });
  }

  /**
   * @returns Type of destroyment
   */
  getDestroyState() {
    return this.destroyType;
  }

  /**
   * @returns If worker pool was destroyed
   * To know if it will be destroyed, @see `getDestroyState()`
   */
  isDestroyed() {
    return this.destroyType == "none" && this.onDestroyCallback === undefined;
  }

  /**
   * Adds new job for workers to execute
   * @param job Job to execute
   * @returns If job has been added to waiting/execution queue
   * Can return false if the WorkerPool is being destroyed
   */
  addJob(job: WorkerJob): boolean {
    if (this.destroyType != "none") return false;

    this.waitingJobs.addBack(job); // Adding job to queue
    this.workerFinished(); // Dispatching potential worker
    return true;
  }
}
