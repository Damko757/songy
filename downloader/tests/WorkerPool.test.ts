import { afterAll, beforeAll, describe, expect, it, jest } from "@jest/globals";
import { WorkerPool } from "../src/Workers/WorkerPool.js";
import { ObjectId } from "mongoose";
import { Worker } from "node:worker_threads";
import { WorkerJob } from "../src/Commands/Command.js";
import path, { dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { sha256 } from "sha.js";
import fs from "fs";
import { Downloader } from "../src/Downloader.js";
import { afterEach } from "node:test";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

class WorkerPoolTest extends WorkerPool {
  protected workerScriptFile() {
    // Should regenerate
    return path.resolve(__dirname, "../dist/downloader/src/Workers/Worker.js");
  }
}

beforeAll(() => {
  Downloader.downloadDirectory = path.resolve(
    __dirname,
    "../dist/downloader/out"
  ); // Temporary directory
});
afterAll(() => {
  // Removing out folder
  try {
    fs.rmSync(Downloader.downloadDirectory, { recursive: true });
  } catch (e) {}
});

describe("Invalid behaviour", () => {
  it("Invalid worker number", () => {
    expect(() => new WorkerPool(0)).toThrow();
    expect(() => new WorkerPool(-1)).toThrow("Invalid number of workers");
    expect(async () => await new WorkerPool(1).destroy("force")).not.toThrow();
  });
});

describe("Initialization", () => {
  for (let i = 1; i <= 15; i++) {
    it(`Finish-all pings with ${i} workers`, () => {
      const wp = new WorkerPoolTest(i);
      const jobs: WorkerJob[] = Array.from(Array(10)).map((_, i) => {
        const job: WorkerJob = {
          job: {
            action: "ping",
            value: i,
          },
          handlers: {
            pong: (msg) => {
              expect(msg.value).toBe(i);
            }, // Checking return value from it
          },
        };
        return job;
      });
      jobs.forEach(wp.addJob, wp); // Adding job

      return wp.destroy("finish-all");
    });
  }

  it("Order check", () => {
    expect.assertions(10);
    const wp = new WorkerPoolTest(1);
    let c = 0; // Correct Order counter
    const jobs: WorkerJob[] = Array.from(Array(10)).map((_, i) => {
      const job: WorkerJob = {
        job: {
          action: "ping",
          value: i,
        },
        handlers: {
          pong: (msg) => {
            expect(msg.value).toBe(c++);
          }, // Checking return value from it
        },
      };
      return job;
    });
    jobs.forEach(wp.addJob, wp); // Adding job

    return wp.destroy("finish-all");
  });
});

describe("Downloading", () => {
  it("3 Concurrent videos", () => {
    const wp = new WorkerPoolTest(3);

    return Promise.all(
      Array.from(Array(3)).map(
        (_, i) =>
          new Promise<void>((resolve) => {
            let downloaded = 0;
            let total = 0;
            const id = "ABCDE"[i] as unknown as ObjectId;
            wp.addJob({
              job: {
                action: "download",
                extension: "mp4",
                id: id,
                link: "https://www.youtube.com/watch?v=ucZl6vQ_8Uo",
                options: {
                  video: { quality: "lowestvideo" },
                  audio: { quality: "lowestaudio" },
                },
                type: "video",
              },
              handlers: {
                progress: (msg) => {
                  downloaded = msg.downloaded;
                  total = msg.total;
                },
                end: async () => {
                  expect(downloaded).toBe(total);
                  expect(total).not.toBe(0);

                  // Checking if temporary files are deleted
                  // They are deleted asyncly, hence the wait
                  await new Promise<void>((resolve) => {
                    setTimeout(() => {
                      expect(
                        fs.existsSync(
                          Downloader.downloadPath(id, "mp3", "audio")
                        )
                      ).toBeFalsy();
                      expect(
                        fs.existsSync(
                          Downloader.downloadPath(id, "mp4", "video")
                        )
                      ).toBeFalsy();
                      resolve();
                    }, 1000);
                  });

                  // Real file kept
                  expect(() =>
                    fs.accessSync(Downloader.downloadPath(id, "mp4"))
                  ).not.toThrow();

                  // TODO fetch from correct directory
                  const hash = new sha256();
                  hash.update(
                    fs.readFileSync(Downloader.downloadPath(id, "mp4"))
                  );
                  expect(hash).toMatchSnapshot();

                  resolve();
                },
              },
            });
          })
      )
    );
  }, 120_000); // 2 min
});
