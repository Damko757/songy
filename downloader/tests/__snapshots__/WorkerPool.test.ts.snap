// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Downloading 3 Concurrent videos 1`] = `
Sha256 {
  "_a": -746243170,
  "_b": 1120693629,
  "_block": {
    "data": [
      0,
      0,
      0,
      0,
      76,
      97,
      118,
      102,
      53,
      57,
      46,
      50,
      55,
      46,
      49,
      48,
      48,
      0,
      0,
      109,
      100,
      105,
      114,
      97,
      112,
      112,
      108,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      45,
      105,
      108,
      115,
      116,
      0,
      0,
      0,
      37,
      169,
      116,
      111,
      111,
      0,
      0,
      0,
      29,
      100,
      97,
      116,
      97,
      0,
      0,
      0,
      1,
    ],
    "type": "Buffer",
  },
  "_blockSize": 64,
  "_c": -311441869,
  "_d": -1626511018,
  "_e": -2037773801,
  "_f": -603391055,
  "_finalSize": 56,
  "_g": 626987033,
  "_h": 1492319933,
  "_len": 746705,
  "_w": [
    0,
    8552,
    1684828672,
    0,
    109,
    1684632161,
    1886415872,
    0,
    0,
    45,
    1768715124,
    37,
    -1451987089,
    29,
    1684108385,
    1,
    1828686717,
    2050131099,
    -1724475865,
    -1773138310,
    751769684,
    -443456025,
    -543421383,
    192279968,
    1542059998,
    -692077360,
    247979750,
    675186254,
    -575042395,
    115187851,
    1774773513,
    -707223038,
    1251221319,
    -938736731,
    843516379,
    1634955055,
    596123291,
    -1794696780,
    -31777425,
    -1877923142,
    1701461720,
    -206447429,
    28170870,
    1424115365,
    1210786005,
    -1758147185,
    634136589,
    -1163047268,
    567094215,
    1254613796,
    1088662396,
    -317486893,
    426132725,
    2004925339,
    -784890593,
    810751643,
    17442330,
    -1194077266,
    405552855,
    1836428146,
    -1760668045,
    -567685810,
    -1250322560,
    1534896820,
  ],
}
`;

exports[`Downloading 3 Concurrent videos 2`] = `
Sha256 {
  "_a": -746243170,
  "_b": 1120693629,
  "_block": {
    "data": [
      0,
      0,
      0,
      0,
      76,
      97,
      118,
      102,
      53,
      57,
      46,
      50,
      55,
      46,
      49,
      48,
      48,
      0,
      0,
      109,
      100,
      105,
      114,
      97,
      112,
      112,
      108,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      45,
      105,
      108,
      115,
      116,
      0,
      0,
      0,
      37,
      169,
      116,
      111,
      111,
      0,
      0,
      0,
      29,
      100,
      97,
      116,
      97,
      0,
      0,
      0,
      1,
    ],
    "type": "Buffer",
  },
  "_blockSize": 64,
  "_c": -311441869,
  "_d": -1626511018,
  "_e": -2037773801,
  "_f": -603391055,
  "_finalSize": 56,
  "_g": 626987033,
  "_h": 1492319933,
  "_len": 746705,
  "_w": [
    0,
    8552,
    1684828672,
    0,
    109,
    1684632161,
    1886415872,
    0,
    0,
    45,
    1768715124,
    37,
    -1451987089,
    29,
    1684108385,
    1,
    1828686717,
    2050131099,
    -1724475865,
    -1773138310,
    751769684,
    -443456025,
    -543421383,
    192279968,
    1542059998,
    -692077360,
    247979750,
    675186254,
    -575042395,
    115187851,
    1774773513,
    -707223038,
    1251221319,
    -938736731,
    843516379,
    1634955055,
    596123291,
    -1794696780,
    -31777425,
    -1877923142,
    1701461720,
    -206447429,
    28170870,
    1424115365,
    1210786005,
    -1758147185,
    634136589,
    -1163047268,
    567094215,
    1254613796,
    1088662396,
    -317486893,
    426132725,
    2004925339,
    -784890593,
    810751643,
    17442330,
    -1194077266,
    405552855,
    1836428146,
    -1760668045,
    -567685810,
    -1250322560,
    1534896820,
  ],
}
`;

exports[`Downloading 3 Concurrent videos 3`] = `
Sha256 {
  "_a": -746243170,
  "_b": 1120693629,
  "_block": {
    "data": [
      0,
      0,
      0,
      0,
      76,
      97,
      118,
      102,
      53,
      57,
      46,
      50,
      55,
      46,
      49,
      48,
      48,
      0,
      0,
      109,
      100,
      105,
      114,
      97,
      112,
      112,
      108,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      45,
      105,
      108,
      115,
      116,
      0,
      0,
      0,
      37,
      169,
      116,
      111,
      111,
      0,
      0,
      0,
      29,
      100,
      97,
      116,
      97,
      0,
      0,
      0,
      1,
    ],
    "type": "Buffer",
  },
  "_blockSize": 64,
  "_c": -311441869,
  "_d": -1626511018,
  "_e": -2037773801,
  "_f": -603391055,
  "_finalSize": 56,
  "_g": 626987033,
  "_h": 1492319933,
  "_len": 746705,
  "_w": [
    0,
    8552,
    1684828672,
    0,
    109,
    1684632161,
    1886415872,
    0,
    0,
    45,
    1768715124,
    37,
    -1451987089,
    29,
    1684108385,
    1,
    1828686717,
    2050131099,
    -1724475865,
    -1773138310,
    751769684,
    -443456025,
    -543421383,
    192279968,
    1542059998,
    -692077360,
    247979750,
    675186254,
    -575042395,
    115187851,
    1774773513,
    -707223038,
    1251221319,
    -938736731,
    843516379,
    1634955055,
    596123291,
    -1794696780,
    -31777425,
    -1877923142,
    1701461720,
    -206447429,
    28170870,
    1424115365,
    1210786005,
    -1758147185,
    634136589,
    -1163047268,
    567094215,
    1254613796,
    1088662396,
    -317486893,
    426132725,
    2004925339,
    -784890593,
    810751643,
    17442330,
    -1194077266,
    405552855,
    1836428146,
    -1760668045,
    -567685810,
    -1250322560,
    1534896820,
  ],
}
`;
